src/helper_functions.rs

```diff
@@ -0,0 +1,206 @@
+use crate::RPSGame;
+use gstd::{exec, msg, prelude::*, ActorId};
+use rps_io::*;
+
+impl RPSGame {
+    pub(crate) fn change_stage_by_timeout_if_needed(&mut self) {
+        let end_time = self.current_stage_start_timestamp
+            + match self.stage {
+                GameStage::Preparation => self.game_config.entry_timeout,
+                GameStage::InProgress(_) => self.game_config.move_timeout,
+                GameStage::Reveal(_) => self.game_config.reveal_timeout,
+            };
+
+        if end_time < exec::block_timestamp() {
+            match &self.stage {
+                GameStage::Preparation => self.handle_preparation_timeout(),
+                GameStage::InProgress(_) => self.handle_moves_timeout(),
+                GameStage::Reveal(description) => {
+                    self.handle_reveal_timeout(description.finished_players.len())
+                }
+            }
+        }
+    }
+
+    pub(crate) fn bytes_to_hex_string(bytes: [u8; 32]) -> String {
+        bytes
+            .iter()
+            .map(|b| format!("{:02x}", b))
+            .collect::<String>()
+    }
+
+    pub(crate) fn transit_to_in_progress_stage_from_preparation(&mut self) {
+        let progress_description = StageDescription {
+            anticipated_players: self.lobby.clone(),
+            finished_players: Default::default(),
+        };
+
+        self.stage = GameStage::InProgress(progress_description);
+        self.update_timestamp();
+    }
+
+    pub(crate) fn try_to_transit_to_reveal_stage_after_move(&mut self) {
+        if let GameStage::InProgress(description) = &self.stage {
+            if description.anticipated_players.is_empty() {
+                self.transit_to_reveal_stage(description.finished_players.clone());
+            }
+        }
+    }
+
+    pub(crate) fn handle_preparation_timeout(&mut self) {
+        match self.lobby.len() {
+            0 | 1 => self.update_timestamp(),
+            _ => self.transit_to_in_progress_stage_from_preparation(),
+        }
+    }
+
+    pub(crate) fn handle_moves_timeout(&mut self) {
+        let finished_players = match &self.stage {
+            GameStage::Preparation | GameStage::Reveal(_) => panic!("Wrong stage"),
+            GameStage::InProgress(description) => &description.finished_players,
+        };
+
+        match finished_players.len() {
+            0 => self.update_timestamp(),
+            1 => {
+                let winner = finished_players
+                    .clone()
+                    .into_iter()
+                    .last()
+                    .expect("Unknown winner");
+                msg::send(winner, "", exec::value_available()).expect("Can't send reward");
```

Comment:
No need to clone the whole BTreeSet:
```suggestion
                let winner = finished_players
                    .iter()
                    .last()
                    .expect("Unknown winner");
                msg::send(*winner, "", exec::value_available()).expect("Can't send reward");
```